File: platform/robot/src/hardware/button/event.py
from enum import StrEnum


class ButtonEvent(StrEnum):
    """Button event types."""

    PRESSED = "pressed"
    RELEASED = "released"
    SHORT_PRESS = "short_press"
    LONG_PRESS = "long_press"

    SHUTDOWN_PRESS = "shutdown_press"
    """A very long hold, meaning "power the robot down cleanly".

    Exists because the only safe shutdown today needs a laptop, a network and
    SSH -- none of which are guaranteed at a competition table, where cutting
    power to a running Zero risks the ext4 corruption safe-shutdown-zero.sh
    documents. This is the same intent as a physical power gesture.

    Never emitted while racing: see the state machine's handler. Between the
    long-press threshold and this one, a hold always means stop, never
    power-off.
    """

File: src/go/internal/driver/button/event.go
package button

// Kind identifies which button event Read returned. Mirrors
// platform/robot/src/hardware/button/event.py's ButtonEvent enum.
type Kind int

// Event is one debounced button transition: a press, a release, or a hold
// crossing one of the two configured thresholds while still held.
type Event struct {
	Kind Kind
	// HeldSec is how long the button had been held, in seconds, at the
	// moment this event fired. Zero for KindPressed.
	HeldSec float64
}

const (
	// KindPressed fires the instant a debounced press begins.
	KindPressed Kind = iota
	// KindLongPress fires once, while still held, when the hold crosses
	// Thresholds.LongPressThreshold.
	KindLongPress
	// KindShutdownPress fires once, while still held, when the hold
	// crosses Thresholds.ShutdownPressThreshold. Never emitted while
	// racing -- see event.py's ButtonEvent.SHUTDOWN_PRESS doc comment;
	// that policy lives in the state machine consumer, not here.
	KindShutdownPress
	// KindShortPress fires on release if neither hold threshold fired
	// during the press.
	KindShortPress
	// KindReleased fires on release if a hold threshold already fired
	// during the press -- the hold event itself was the actionable
	// signal, so release only clears it rather than firing a second
	// action for the same gesture.
	KindReleased
)

// String returns the wire-style spelling used by the Python driver this
// ports (platform/robot/src/hardware/button/event.py), e.g. "short_press".
func (k Kind) String() string {
	switch k {
	case KindPressed:
		return "pressed"
	case KindLongPress:
		return "long_press"
	case KindShutdownPress:
		return "shutdown_press"
	case KindShortPress:
		return "short_press"
	case KindReleased:
		return "released"
	default:
		return "unknown"
	}
}
